src/gluonts/dataset/rolling_dataset.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from functools import partial
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+from gluonts.dataset.common import Dataset
+from gluonts.dataset.util import to_pandas
+from pydantic import BaseModel
+
+
+class StepStrategy(BaseModel):
+    """
+    Removes datapoints equivalent to step_size for each iteration until
+    amount of data left is less than prediction_length
+
+    Parameters
+    ----------
+    prediction_length
+        The prediction length of the Predictor that the dataset will be
+        used with
+
+    Returns
+    -------
+    A partial function which yields the rolled windows
+    """
+
+    prediction_length: int
+    step_size: int = 1
+
+    def get_windows(self, window):
+        assert (
+            self.prediction_length > 0
+        ), """the step strategy requires a prediction_length > 0"""
+        assert self.step_size > 0, """step_size should be > 0"""
+
+        while len(window) >= self.prediction_length:
+            yield window
+            window = window[: -self.step_size]
+
+
+class NumSplitsStrategy(BaseModel):
+    prediction_length: int
+    num_splits: int
+
+    def get_windows(self, window):
+        assert num_splits > 1, """num_splits should be > 1"""
+        for slice_idx in np.linspace(
+            start=self.prediction_length, stop=len(window), num=self.num_splits
+        ):
+
+            yield window[: int(round(slice_idx))]
+
+
+def generate_rolling_datasets(
+    dataset: Dataset,
+    strategy,
+    start_time: pd.Timestamp,
+    end_time: Optional[pd.Timestamp] = None,
+) -> Dataset:
+    """
+    Parameters
+    ----------
+    dataset
+        Dataset to generate the rolling forecasting datasets from
+    strategy
+        The strategy that is to be used when rolling
+    start_time
+        The start of the window where rolling forecasts should be applied
+    end_time
+        The end time of the window where rolling should be applied
+    Returns
+    -------
+    Dataset
+        The augmented dataset
```

Comment:
I think the convention is that these go to the bottom of the docstring

Reply:
This contradicts what @jaheba reviewed on the 8th of June, the examples I found online also has them on the bottom but im not sure. @jaheba what do you think?

Reply:
That's fine, as long as it displays nicely when API docs are generated, I can take a look how sphinx reads it

Reply:
it looks like this is displayed weirdly:

![image](https://user-images.githubusercontent.com/433963/98812111-3022b100-2422-11eb-9735-9086e590c672.png)

Reply:
Updated the formatting in the latest commit, it looks nice on my machine currently :)